generator/template.go

```diff
@@ -195,6 +195,97 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
     }
   })
 }
+
+type RequestPayload = Record<string, unknown>;
+type Primitive = string | boolean | number;
+
+/**
+ * Checks if given value is of a primitive type
+ * @param  {unknown} value
+ * @return {boolean}
+ */
+function isPrimitive(value: unknown): boolean {
+  return (
+    typeof value === "string" ||
+    typeof value === "number" ||
+    typeof value === "boolean"
+  );
+}
+
+/**
+ * Flattens a deeply nested request payload and returns an object
+ * with only primitive values and non-empty array of primitive values
+ * @param  {RequestPayload} requestPayload
+ * @param  {String} path
+ * @return {Record<string, Primitive | Array<Primitive>}
+ */
+function flattenRequestPayload<T extends RequestPayload>(
+  requestPayload: T,
+  path: string = ""
+): Record<string, Primitive | Array<Primitive>> {
+  return Object.keys(requestPayload).reduce(
+    (acc: T, key: string): T => {
+      const value = requestPayload[key];
+      const newPath = path ? [path, key].join(".") : key;
+
+      const isObject =
+        Object.prototype.toString.call(value) === "[object Object]";
+      const isNonEmptyPrimitiveArray =
+        Array.isArray(value) &&
+        value.every(v => isPrimitive(v)) &&
+        value.length > 0;
+
+      switch (true) {
+        case isObject:
+          acc = {
+            ...acc,
+            ...flattenRequestPayload(value as RequestPayload, newPath)
+          };
+          break;
+        case isPrimitive(value) || isNonEmptyPrimitiveArray:
+          acc = { ...acc, [newPath]: value };
+          break;
+      }
+
+      return acc;
+    },
+    {} as T
+  ) as Record<string, Primitive | Array<Primitive>>;
+}
+
+/**
+ * Renders a deeply nested request payload into a string of URL search
+ * parameters by first flattening the request payload and then removing keys
+ * which are already present in the URL path.
+ * @param  {RequestPayload} requestPayload
+ * @param  {string[]} urlPathParams
+ * @return {string}
+ */
+export function renderURLSearchParams<T extends RequestPayload>(
+  requestPayload: T,
+  urlPathParams: string[] = []
+): string {
+  const flattenedRequestPayload = flattenRequestPayload(requestPayload);
+
+  const urlSearchParams = Object.keys(flattenedRequestPayload).reduce(
+    (acc: string[][], key: string): string[][] => {
+      // key should not be present in the url path as a parameter
+      const value = flattenedRequestPayload[key];
+      if (!urlPathParams.find(f => f === key)) {
```

Comment:
can we flatten the nested if like the following?

```
if (!urlPathParams.find(f => f === key)) {
  return acc
}

return Array.isArray(value)
? ...
: ...

```

Reply:
Ah yes, I was just thinking to refactor it that way. Thank you for your suggestion 🙇🏼

Reply:
can we revert the logic here? it's just easier to read without the negative conditions in if.

```
if (urlPathParams.find(f => f === key)) {
  return acc
}
```